File: Harvest/Views/Subscription/PurchaseSheet.swift
import SwiftUI
import StoreKit

struct PurchaseSheet: View {
    let tier: SubscriptionTier
    let viewModel: SubscriptionViewModel
    let authViewModel: AuthViewModel
    @Binding var billingPeriod: BillingPeriod
    @Environment(\.dismiss) private var dismiss

    private var selectedProduct: Product? {
        viewModel.getProduct(for: tier, billingPeriod: billingPeriod)
    }

    private var hasAnyStoreKitProducts: Bool {
        !viewModel.products.isEmpty
    }

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(spacing: HarvestTheme.Spacing.xl) {
                    // Header
                    VStack(spacing: HarvestTheme.Spacing.md) {
                        Image(systemName: tierIcon)
                            .font(.system(size: 60))
                            .foregroundStyle(tierColor)

                        Text("Upgrade to \(tier.marketingDisplayName)")
                            .font(HarvestTheme.Typography.h2)

                        Text(tier.description)
                            .font(HarvestTheme.Typography.bodyRegular)
                            .foregroundStyle(HarvestTheme.Colors.textSecondary)
                            .multilineTextAlignment(.center)
                    }
                    .padding(.top)

                    // Billing Period Selector
                    Picker("Billing Period", selection: $billingPeriod) {
                        Text("Weekly").tag(BillingPeriod.weekly)
                        Text("Monthly").tag(BillingPeriod.monthly)
                    }
                    .pickerStyle(.segmented)
                    .padding(.horizontal)

                    // Price Display
                    if let product = selectedProduct {
                        VStack(spacing: HarvestTheme.Spacing.sm) {
                            Text(product.displayPrice)
                                .font(.system(size: 48, weight: .bold))
                                .foregroundStyle(HarvestTheme.Colors.textPrimary)

                            Text(billingPeriod == .weekly ? "per week" : "per month")
                                .font(HarvestTheme.Typography.bodyRegular)
                                .foregroundStyle(HarvestTheme.Colors.textSecondary)
                        }
                        .padding(.vertical)
                    }

                    // Features
                    GlassCard {
                        VStack(alignment: .leading, spacing: HarvestTheme.Spacing.md) {
                            Text("What you get:")
                                .font(HarvestTheme.Typography.h3)

                            VStack(alignment: .leading, spacing: HarvestTheme.Spacing.sm) {
                                featureRow("Matches per week", value: tier.matchesPerWeek.map { "\($0)" } ?? "Unlimited")
                                featureRow("Search distance", value: tier.maxDistanceMiles.map { "\($0) miles" } ?? "Unlimited")
                                featureRow("Gardener chats/day", value: tier.gardenerConversationsPerDay.map { "\($0)" } ?? "Unlimited")
                                featureRow("Gardener character limit", value: "\(tier.gardenerCharacterLimit / 1000)k")

                                if tier.hasValuesMatching {
                                    featureCheck("Deep values-based matching")
                                }
                                if tier.hasAdvancedFilters {
                                    featureCheck("Advanced discovery filters")
                                }
                                if tier.hasFullFilters {
                                    featureCheck("Complete filter control")
                                }
                                if tier.canSeeLikes {
                                    featureCheck("See who likes you")
                                }
                                if tier.canDisableMindfulMessaging {
                                    featureCheck("Optional mindful messaging")
                                }
                            }
                        }
                    }
                    .padding(.horizontal)

                    // Purchase Button
                    if let product = selectedProduct {
                        GlassButton(
                            title: viewModel.isPurchasing ? "Processing..." : "Subscribe Now",
                            style: .primary
                        ) {
                            Task {
                                if let userId = authViewModel.currentUserId {
                                    await viewModel.purchase(product: product, userId: userId)
                                    if viewModel.error == nil {
                                        dismiss()
                                    }
                                }
                            }
                        }
                        .disabled(viewModel.isPurchasing)
                        .padding(.horizontal)
                    } else if hasAnyStoreKitProducts {
                        VStack(spacing: HarvestTheme.Spacing.sm) {
                            Text("This subscription product is not available right now.")
                                .foregroundStyle(HarvestTheme.Colors.textSecondary)
                                .multilineTextAlignment(.center)

                            Button("Retry") {
                                Task {
                                    await viewModel.loadProducts()
                                }
                            }
                            .font(HarvestTheme.Typography.bodySmall)
                        }
                        .padding(.horizontal)
                    } else {
                        VStack(spacing: HarvestTheme.Spacing.sm) {
                            Text(viewModel.error ?? "Loading product information...")
                                .foregroundStyle(HarvestTheme.Colors.textSecondary)
                                .multilineTextAlignment(.center)

                            Button("Retry") {
                                Task {
                                    await viewModel.loadProducts()
                                }
                            }
                            .font(HarvestTheme.Typography.bodySmall)
                        }
                    }

                    // Fine Print
                    VStack(spacing: HarvestTheme.Spacing.xs) {
                        Text("Payment will be charged to your Apple ID account at confirmation of purchase.")
                            .font(HarvestTheme.Typography.caption)
                            .foregroundStyle(HarvestTheme.Colors.textTertiary)
                            .multilineTextAlignment(.center)

                        Text("Subscription automatically renews unless canceled at least 24 hours before the end of the current period.")
                            .font(HarvestTheme.Typography.caption)
                            .foregroundStyle(HarvestTheme.Colors.textTertiary)
                            .multilineTextAlignment(.center)
                    }
                    .padding(.horizontal, HarvestTheme.Spacing.xl)
                    .padding(.bottom)
                }
            }
            .foregroundStyle(HarvestTheme.Colors.textPrimary)
            .background(HarvestTheme.Colors.background.ignoresSafeArea())
            .navigationTitle("Subscribe")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .topBarTrailing) {
                    Button("Cancel") {
                        dismiss()
                    }
                }
            }
            .toolbarBackground(HarvestTheme.Colors.background, for: .navigationBar)
            .toolbarBackground(.visible, for: .navigationBar)
            .toolbarColorScheme(.dark, for: .navigationBar)
            .task {
                if viewModel.products.isEmpty {
                    await viewModel.loadProducts()
                }
            }
        }
    }

    private var tierIcon: String {
        switch tier.name {
        case .seed: return "leaf"
        case .green: return "leaf.fill"
        case .gold: return "crown.fill"
        }
    }

    private var tierColor: Color {
        switch tier.name {
        case .seed: return HarvestTheme.Colors.textPrimary
        case .green: return HarvestTheme.Colors.accent
        case .gold: return Color(hex: "F59E0B")
        }
    }

    private func featureRow(_ label: String, value: String) -> some View {
        HStack {
            Text(label)
                .font(HarvestTheme.Typography.bodySmall)
                .foregroundStyle(HarvestTheme.Colors.textSecondary)
            Spacer()
            Text(value)
                .font(HarvestTheme.Typography.bodySmall)
                .fontWeight(.semibold)
                .foregroundStyle(HarvestTheme.Colors.textPrimary)
        }
    }

    private func featureCheck(_ label: String) -> some View {
        HStack(spacing: HarvestTheme.Spacing.sm) {
            Image(systemName: "checkmark.circle.fill")
                .foregroundStyle(HarvestTheme.Colors.accent)
            Text(label)
                .font(HarvestTheme.Typography.bodySmall)
                .foregroundStyle(HarvestTheme.Colors.textPrimary)
            Spacer()
        }
    }
}

File: android/app/src/main/java/com/harvestglass/harvest/ui/subscription/PurchaseSheet.kt
package com.harvestglass.harvest.ui.subscription

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.border
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.android.billingclient.api.ProductDetails
import com.harvestglass.harvest.billing.BillingPeriod
import com.harvestglass.harvest.data.model.FieldFilterLevel
import com.harvestglass.harvest.data.model.SubscriptionTier
import com.harvestglass.harvest.ui.components.GlassCard
import com.harvestglass.harvest.ui.components.GlassCardStyle
import com.harvestglass.harvest.ui.theme.HarvestButton
import com.harvestglass.harvest.ui.theme.HarvestButtonKind
import com.harvestglass.harvest.ui.theme.HarvestTheme

/**
 * Port of Harvest/Views/Subscription/PurchaseSheet.swift.
 *
 * Prices come from Play rather than the tier table — Google localises and
 * converts them, and the store's number is the one the user is charged.
 */
@Composable
fun PurchaseSheet(
    tier: SubscriptionTier,
    userId: String,
    state: SubscriptionUiState,
    billingPeriod: BillingPeriod,
    onBillingPeriodChange: (BillingPeriod) -> Unit,
    onRetryProducts: () -> Unit,
    onSubscribe: (ProductDetails) -> Unit,
    onDismiss: () -> Unit
) {
    val selected = state.product(tier, billingPeriod)

    LaunchedEffect(userId) {
        if (state.products.isEmpty()) onRetryProducts()
    }

    Column(
        Modifier
            .fillMaxSize()
            .background(HarvestTheme.Colors.formBackground)
    ) {
        PurchaseSheetTopBar(onDismiss)

        Column(
            verticalArrangement = Arrangement.spacedBy(HarvestTheme.Spacing.xl),
            horizontalAlignment = Alignment.CenterHorizontally,
            modifier = Modifier
                .weight(1f)
                .verticalScroll(rememberScrollState())
                .padding(HarvestTheme.Spacing.md)
                .navigationBarsPadding()
        ) {
            Column(
                verticalArrangement = Arrangement.spacedBy(HarvestTheme.Spacing.md),
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                Icon(
                    imageVector = tier.name.icon(),
                    contentDescription = null,
                    tint = tier.name.color(),
                    modifier = Modifier.size(60.dp)
                )
                Text(
                    text = "Upgrade to ${tier.marketingDisplayName}",
                    style = HarvestTheme.Typography.h2,
                    color = HarvestTheme.Colors.textPrimary,
                    textAlign = TextAlign.Center
                )
                tier.description?.let {
                    Text(
                        text = it,
                        style = HarvestTheme.Typography.bodyRegular,
                        color = HarvestTheme.Colors.textSecondary,
                        textAlign = TextAlign.Center
                    )
                }
            }

            BillingPeriodPicker(billingPeriod, onBillingPeriodChange)

            selected?.let { details ->
                Column(
                    verticalArrangement = Arrangement.spacedBy(HarvestTheme.Spacing.sm),
                    horizontalAlignment = Alignment.CenterHorizontally
                ) {
                    Text(
                        text = details.formattedPrice(),
                        style = HarvestTheme.Typography.display.copy(fontSize = 48.sp),
                        color = HarvestTheme.Colors.textPrimary
                    )
                    Text(
                        text = if (billingPeriod == BillingPeriod.WEEKLY) "per week" else "per month",
                        style = HarvestTheme.Typography.bodyRegular,
                        color = HarvestTheme.Colors.textSecondary
                    )
                }
            }

            GlassCard(style = GlassCardStyle.LIGHT) {
                Column(verticalArrangement = Arrangement.spacedBy(HarvestTheme.Spacing.md)) {
                    Text(
                        text = "What you get:",
                        style = HarvestTheme.Typography.h3,
                        color = HarvestTheme.Colors.textPrimary
                    )
                    Column(verticalArrangement = Arrangement.spacedBy(HarvestTheme.Spacing.sm)) {
                        FeatureRow("Seeds per day", "${tier.dailySeedLimit}")
                        FeatureRow("Receive Seeds", "Unlimited")
                        FeatureRow(
                            "Gardener chat",
                            "${tier.gardenerCharacterLimit.grouped()} characters/day"
                        )
                        FeatureRow("Screenshot reviews", "${tier.gardenerScreenshotsPerDay}/day")

                        when (tier.fieldFilterLevel) {
                            FieldFilterLevel.NONE -> Unit
                            FieldFilterLevel.ADVANCED -> IncludedRow("Advanced room member filters")
                            FieldFilterLevel.FULL -> IncludedRow("Every room member filter")
                        }
                        if (tier.hasDeepSoilInsights) {
                            IncludedRow("Deeper Soil & value insights")
                            IncludedRow("Advanced compatibility insights")
                        }
                        if (tier.hasGrowthFeatures) {
                            IncludedRow("Premium relationship growth features")
                        }
                    }
                }
            }

            if (selected != null) {
                HarvestButton(
                    text = if (state.isPurchasing) "Processing..." else "Subscribe Now",
                    kind = HarvestButtonKind.PRIMARY,
                    modifier = Modifier.fillMaxWidth()
                ) {
                    if (!state.isPurchasing) onSubscribe(selected)
                }
            } else {
                Column(
                    verticalArrangement = Arrangement.spacedBy(HarvestTheme.Spacing.sm),
                    horizontalAlignment = Alignment.CenterHorizontally
                ) {
                    Text(
                        text = if (state.products.isNotEmpty()) {
                            "This subscription product is not available right now."
                        } else {
                            state.error ?: "Loading product information..."
                        },
                        style = HarvestTheme.Typography.bodyRegular,
                        color = HarvestTheme.Colors.textSecondary,
                        textAlign = TextAlign.Center
                    )
                    Text(
                        text = "Retry",
                        style = HarvestTheme.Typography.bodySmall,
                        color = HarvestTheme.Colors.accent,
                        modifier = Modifier.clickable { onRetryProducts() }
                    )
                }
            }

            Column(
                verticalArrangement = Arrangement.spacedBy(HarvestTheme.Spacing.xs),
                horizontalAlignment = Alignment.CenterHorizontally,
                modifier = Modifier.padding(horizontal = HarvestTheme.Spacing.xl)
            ) {
                Text(
                    text = "Payment will be charged to your Google Play account at " +
                        "confirmation of purchase.",
                    style = HarvestTheme.Typography.caption,
                    color = HarvestTheme.Colors.textTertiary,
                    textAlign = TextAlign.Center
                )
                Text(
                    text = "Subscription automatically renews unless canceled at least 24 " +
                        "hours before the end of the current period.",
                    style = HarvestTheme.Typography.caption,
                    color = HarvestTheme.Colors.textTertiary,
                    textAlign = TextAlign.Center
                )
            }
        }
    }
}

@Composable
private fun PurchaseSheetTopBar(onDismiss: () -> Unit) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        modifier = Modifier
            .fillMaxWidth()
            .background(HarvestTheme.Colors.wineBlack)
            .statusBarsPadding()
            .padding(HarvestTheme.Spacing.md)
    ) {
        Text(
            text = "Subscribe",
            style = HarvestTheme.Typography.h4,
            color = HarvestTheme.Colors.textPrimary,
            modifier = Modifier.weight(1f)
        )
        Text(
            text = "Cancel",
            style = HarvestTheme.Typography.bodyRegular,
            color = HarvestTheme.Colors.accent,
            modifier = Modifier.clickable { onDismiss() }
        )
    }
}

/** Stands in for the iOS segmented Picker, matching the pair used elsewhere. */
@Composable
private fun BillingPeriodPicker(
    selected: BillingPeriod,
    onSelect: (BillingPeriod) -> Unit
) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(HarvestTheme.Spacing.sm),
        modifier = Modifier.fillMaxWidth()
    ) {
        BillingPeriod.entries.forEach { period ->
            val isSelected = period == selected
            val shape = RoundedCornerShape(HarvestTheme.Radius.sm)
            Text(
                text = period.label(),
                style = HarvestTheme.Typography.bodySmall,
                fontWeight = if (isSelected) FontWeight.SemiBold else FontWeight.Normal,
                color = if (isSelected) {
                    HarvestTheme.Colors.textOnRedPrimary
                } else {
                    HarvestTheme.Colors.textPrimary
                },
                textAlign = TextAlign.Center,
                modifier = Modifier
                    .weight(1f)
                    .background(
                        if (isSelected) {
                            HarvestTheme.Colors.primary
                        } else {
                            HarvestTheme.Colors.formSurface
                        },
                        shape
                    )
                    .border(1.dp, HarvestTheme.Colors.formBorder, shape)
                    .clickable { onSelect(period) }
                    .padding(vertical = HarvestTheme.Spacing.sm)
            )
        }
    }
}

@Composable
private fun IncludedRow(label: String) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(HarvestTheme.Spacing.sm),
        modifier = Modifier.fillMaxWidth()
    ) {
        Icon(
            Icons.Filled.CheckCircle,
            contentDescription = null,
            tint = HarvestTheme.Colors.accent,
            modifier = Modifier.size(18.dp)
        )
        Text(
            text = label,
            style = HarvestTheme.Typography.bodySmall,
            color = HarvestTheme.Colors.textPrimary
        )
    }
}

private fun BillingPeriod.label(): String =
    if (this == BillingPeriod.WEEKLY) "Weekly" else "Monthly"

/**
 * The recurring price Play quotes for this product.
 *
 * A base plan can carry an intro phase, so the last phase is the one that
 * actually recurs — that's the number the sheet's "per week/month" describes.
 */
private fun ProductDetails.formattedPrice(): String =
    subscriptionOfferDetails
        ?.firstOrNull()
        ?.pricingPhases
        ?.pricingPhaseList
        ?.lastOrNull()
        ?.formattedPrice
        .orEmpty()
